build.sbt

```diff
@@ -397,14 +397,11 @@ lazy val compilerInterface212 = (project in internalPath / "compiler-interface")
       import com.typesafe.tools.mima.core._
       import com.typesafe.tools.mima.core.ProblemFilters._
       Seq(
-        exclude[ReversedMissingMethodProblem]("xsbti.compile.ExternalHooks#Lookup.hashClasspath"),
-        exclude[ReversedMissingMethodProblem]("xsbti.compile.ScalaInstance.loaderLibraryOnly"),
-        exclude[ReversedMissingMethodProblem]("xsbti.compile.ScalaInstance.libraryJars"), // Added in 1.3.0
-        exclude[DirectMissingMethodProblem]("xsbti.api.AnalyzedClass.of"),
-        exclude[DirectMissingMethodProblem]("xsbti.api.AnalyzedClass.create"),
-        exclude[ReversedMissingMethodProblem]("xsbti.AnalysisCallback.classesInOutputJar"),
-        exclude[ReversedMissingMethodProblem]("xsbti.compile.IncrementalCompiler.compile"),
-        exclude[DirectMissingMethodProblem]("xsbti.compile.IncrementalCompiler.compile")
+        exclude[IncompatibleMethTypeProblem]("xsbti.*"),
+        exclude[ReversedMissingMethodProblem]("xsbti.*"),
+        exclude[DirectMissingMethodProblem]("xsbti.*"),
+        exclude[IncompatibleSignatureProblem]("xsbti.*"),
+        exclude[IncompatibleResultTypeProblem]("xsbti.*"),
```

Comment:
Ouch.  I'm fairly sure I've relied on these as public API somewhere.  I think I even raised as an action point to Jorge when he was helping define the public API of Zinc to decide (and document) if these are public APIs.

Maybe at the very least it's worth enumerating which API points break with this?  (i.e. Expand the `*` glob in the xsbti non-internal excludes.)

Reply:
compiler-interface as a whole is a contract between Zinc and Scala compiler and it should not be considered Zinc's API. Some datatypes might be part of Zinc API, but the whole thing is not. It would be good to make distinction between the two.

Reply:
Bring back mima here - b8cab8a. We might also side step these issues by moving to a new package too.

Reply:
> We might also side step these issues by moving to a new package too.

I wish this had been implemented. Dotty depends on compiler-interface so this effectively breaks every single Dotty version until we update (at which point there will be some Dotty versions that only work on the latest sbt and some that only work on a previous one). At the very least, it would have been nice to get a ping here so we could start planning for this (I found this randomly).

Reply:
Is there a timeline for sbt/zinc 1.4 ? That would give me some idea of when I need to switch dotty over to the new compiler-interface.

Reply:
@smarter Sorry that I hadn't communicated to Dotty team directly.

There is still an on-going effort to move the compiler-interface out to https://github.com/scala/compiler-interface/ and Scala Center is working to make the bridge less dependent on either 2.x or 3.x compiler internals (https://github.com/scala/compiler-interface/pull/16). I'm still hoping that the new interface will be under different package name once things are settled (https://github.com/scala/compiler-interface/pull/15). If things work out SC may just give you a new bridge? Let's say we aim for some definition of "summer", but support Dotty would be a blocker for final version.

Reply:
> If things work out SC may just give you a new bridge?

IMO, the bridge being in dotty is the best place for it: if it's moved out we cannot add a new feature to dotty with proper incremental compilation support without coordinating a new release of the bridge.

> Let's say we aim for some definition of "summer", but support Dotty would be a blocker for final version.

Thanks, that's good to know!

Reply:
Yes. Once the minimum contract between Zinc and compiler(s) are made, the bridge part would be in respective compiler repo.

Reply:
So for the zinc 1.4 release:
  * the compiler interface will be shipped separately
  * changes to the compiler interface (like VirtualFile) will be versioned somehow (eg in the package name?)
  * zinc 1.4 will support multiple versions of the compiler interface, so zinc 1.4 will work with existing dotty releases, but also with new dotty releases that ship a new compiler bridge that implements the new compiler interface

Is that the plan?

Reply:
- the compiler interface will be shipped separately
- the initial version in a new package will make a big bang change to support VirtualFile, pipelining, and SAX-like API
- Zinc 1.4 supports will support only the new compiler interface. Build tools + Scalac implementations can provide Maven coordinates for a compiler bridge that implements the new compiler interface.
- for Scala ~2.13.2, sbt/zinc can continue to provide source-based compiler bridge.
- for Dotty, maybe there should be a cross publishing of bridges during the transition period like scala-library-compat?

Reply:
> for Dotty, maybe there should be a cross publishing of bridges during the transition period like scala-library-compat?

Not something I want to maintain (especially since we'd have to do this for every single released dotty version which uses a different bridge!). If sbt breaks stuff, then we'll require the new sbt for new dotty version and the old sbt for old dotty versions, it's unfortunate but anything more complicated will just make me burn out of dealing with sbt forever.

Reply:
I think the "new" compiler interface should be the old interface + the new alternative branches (with VirtualFile, SAX, etc), perhaps in a new package for clarity, and Zinc 1.4 should try for the new and fallback to the old, meaning sbt-dotty can upgrade in its own time.

Reply:
Spun out into a ticket, if only to track it for the release: https://github.com/sbt/zinc/issues/779.

Reply:
I just remembered that there's already a PR that implements a versioning mechanism for the compiler-interface: https://github.com/sbt/zinc/pull/661, could it be used here?

Reply:
If there is a way to move forward without breaking the old xsbti api, I'd strongly suggest that be done. Breaking binary compatibility causes downstream pain: https://github.com/sbt/sbt/issues/5532. I took a quick look at mill and it will definitely not work with zinc 1.4.x right now because it passes `java.io.File`s into incremental compiler.

Reply:
I don't know when/if https://github.com/scala/compiler-interface is going to materialize, but I've brought back some compat to xsbti.* classes so Dotty is not broken - https://github.com/sbt/zinc/pull/829